drafts/arcalinea-tx-expiry/draft.rst

```diff
@@ -0,0 +1,61 @@
+**Title:** Transaction Expiry
+
+**Author:** Jay Graber
+
+**Status:** Active
+
+**Category:** Standards
+
+**Created:** 2018-01-09
+
+Abstract
+===========
+
+This is an Standards ZIP describing a new consensus rule to expire transactions that are not mined from the mempool.
+
+Motivation
+===========
+
+Transactions that are too large or have insufficient fees are often not mined. This indeterminism is a source of confusion for users and wallets. Giving users the ability to specify the number of blocks a transaction should sit in the mempool before it expires, and otherwise setting a default, would provide certainty around how long a transaction has to confirm.
+
+Advantages include improving performance by removing transactions unlikely to be mined, and simplifying bidirectional payment channels by reducing the need to store and compress revocations for past states since transactions are allowed to expire.
+
+Specification
+===============
+
+Transactions will have a new field, nBlockExpiry, which will set the blockheight after which transactions will be removed from the mempool if they have not been mined. The last block that the transaction below could possibly be included in is 3539. After that, it will be removed from the mempool.
+
+````
+"txid": "17561b98cc77cd5a984bb959203e073b5f33cf14cbce90eb32b95ae2c796723f",
+"version": 3,
+"locktime": 2089,
+"blockexpiry": 3539,
+```
+
+With the addition of this feature, zero-confirmation transactions with an expiration blockheight set will have even less guarantee of inclusion. This means that UIs and services must never rely on zero-confirmation transactions in Zcash.
+
+Default behavior
+-----------------
+
+Default behavior does not change -- transactions do not expire. (Alternative proposal is to set a default expiry period of 2 days, ~1440 blocks.) When nBlockExpiry is set to 0, the transaction will not expire from the mempool.
```

Comment:
Hmm... I'm kinda keen to follow the `nSequence` model here, where `UINT_MAX` is the default case, and anything lower is considered a valid block expiry. In particular, I'm thinking of how `nSequence` and `nLockTime` interact (but there isn't a split-range in `nSequence` like there is in `nLockTime`).

Reply:
If we make it `nExpiryTime` and introduce a split-range like in `nLockTime`, would this make a difference?  Higher numbers would be timestamps. And for `nLockTime`, 0 means no locktime, so it would be the same semantics for `nExpiryTime`. 

Another consideration is minimum expiration time. There should probably be a minimum expiration block height, which is something like chainActive.height + 720 (~1 day worth of blocks) to avoid users spamming the network with soon-to-be-expired txs. Some discussion on this PR rejected from Bitcoin for reorg issues: https://github.com/bitcoin/bitcoin/pull/3509
(Think we already face some of these reorg issues due to the way joinsplits are anchored, so can address the same way, by invalidating all dependent txs for a tx which gets dropped from chain upon reorg?)